src/tests/test_first.cpp

```diff
@@ -0,0 +1,87 @@
+//                   ReactivePlusPlus library
+//
+//           Copyright Aleksey Loginov 2022 - present.
+//                     TC Wang 2022 - present.
+//  Distributed under the Boost Software License, Version 1.0.
+//     (See accompanying file LICENSE_1_0.txt or copy at
+//           https://www.boost.org/LICENSE_1_0.txt)
+//
+//  Project home: https://github.com/victimsnino/ReactivePlusPlus
+
+#include <catch2/catch_test_macros.hpp>
+
+#include <vector>
+
+#include <rpp/observables/dynamic_observable.hpp>
+#include <rpp/operators/first.hpp>
+#include <rpp/sources/empty.hpp>
+#include <rpp/sources/error.hpp>
+#include <rpp/sources/just.hpp>
+#include <rpp/sources/never.hpp>
+
+#include "mock_observer.hpp"
+
+SCENARIO("first only emits once", "[first]")
+{
+    GIVEN("observable of -1-2-3-|")
+    {
+        auto mock = mock_observer<int>{};
+        auto obs = rpp::source::just(1,2,3);
+        obs.first()
+            .subscribe(mock);
+        THEN("shall see -1-|")
+        {
+            CHECK(mock.get_received_values() == std::vector{1});
+            CHECK(mock.get_on_completed_count() == 1);
+            CHECK(mock.get_on_error_count() == 0);
+        }
+    }
+
+    GIVEN("observable of never")
+    {
+        auto mock = mock_observer<int>{};
+        auto obs = rpp::source::never<int>();
+        obs.first()
+            .subscribe(mock);
+        THEN("shall see ---|")
+        {
+            CHECK(mock.get_received_values().empty());
+            CHECK(mock.get_on_completed_count() == 0);
+            CHECK(mock.get_on_error_count() == 0);
+        }
+    }
+}
+
+SCENARIO("first forwards error", "[first]")
+{
+    GIVEN("observable of x-|")
+    {
+        auto mock = mock_observer<int>{};
+        auto obs = rpp::source::error<int>(std::make_exception_ptr(std::runtime_error{""}));
+        obs.first()
+            .subscribe(mock);
+        THEN("shall see error and no-completed")
+        {
+            CHECK(mock.get_received_values().empty());
+            CHECK(mock.get_on_completed_count() == 0);
+            CHECK(mock.get_on_error_count() == 1);
+        }
+    }
+}
+
+SCENARIO("first regards completed", "[first]")
+{
+    GIVEN("observable of ---|")
+    {
+        auto mock = mock_observer<int>{};
+        auto obs = rpp::source::empty<int>();
+        obs.first()
+            .subscribe(mock);
+        THEN("shall see ---|")
```

Comment:
Shall see on_error with something like “No items”

Reply:
![image](https://user-images.githubusercontent.com/6766554/185753521-be238a99-1a56-4906-955b-681f3cc04f16.png)

I feel that's the requirement for `Single` source.

Reply:
Rxjava has next one

> The takeFirst operator behaves similarly to first, with the exception of how these operators behave wihen the source Observable emits no items that satisfy the predicate. In such a case, first will throw a NoSuchElementException while takeFirst will return an empty Observable (one that calls onCompleted but never calls onNext).

Reply:
rxcpp also follows this one

Reply:
Also it makes more sense. If I want to get first IF it appears, I can do take(1), but if I want to see first and get error in case of empty source, then first makes sense.

Reply:
That makes sense. I'll fix it.